src/transformers/models/pegasus_x/modeling_pegasus_x.py

```diff
@@ -0,0 +1,1659 @@
+        hidden_states = residual + hidden_states
+
+        global_hidden_states = nn.functional.dropout(global_hidden_states, p=self.dropout, training=self.training)
+        global_hidden_states = global_residual + global_hidden_states
+
+        residual = hidden_states
+        hidden_states = self.final_layer_norm(hidden_states)
+        hidden_states = self.activation_fn(self.fc1(hidden_states))
+        hidden_states = nn.functional.dropout(hidden_states, p=self.activation_dropout, training=self.training)
+        hidden_states = self.fc2(hidden_states)
+        hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
+        hidden_states = residual + hidden_states
+
+        global_residual = global_hidden_states
+        global_hidden_states = self.final_layer_norm(global_hidden_states)
+        global_hidden_states = self.activation_fn(self.fc1(global_hidden_states))
+        global_hidden_states = nn.functional.dropout(
+            global_hidden_states, p=self.activation_dropout, training=self.training
+        )
+        global_hidden_states = self.fc2(global_hidden_states)
+        global_hidden_states = nn.functional.dropout(global_hidden_states, p=self.dropout, training=self.training)
+        global_hidden_states = global_residual + global_hidden_states
+
+        if hidden_states.dtype == torch.float16:
+            if torch.isinf(hidden_states).any() or torch.isnan(hidden_states).any():
+                clamp_value = torch.finfo(hidden_states.dtype).max - 1000
+                hidden_states = torch.clamp(hidden_states, min=-clamp_value, max=clamp_value)
+            if torch.isinf(global_hidden_states).any() or torch.isnan(global_hidden_states).any():
+                clamp_value = torch.finfo(global_hidden_states.dtype).max - 1000
+                global_hidden_states = torch.clamp(global_hidden_states, min=-clamp_value, max=clamp_value)
+
+        outputs = (hidden_states, global_hidden_states)
+
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs
+
+    @classmethod
+    def pad_local_tokens(cls, hidden_states, attention_mask, block_size):
+        # hidden_states: [batch_size, seq_len, hidden_dim]
+        pad_size = block_size // 2
+        mask_min_value = torch.finfo(hidden_states.dtype).min
+        padded_hidden_states = torch.nn.functional.pad(
+            hidden_states,
+            pad=(0, 0, pad_size, pad_size),
+        )
+        padded_mask = torch.nn.functional.pad(
+            attention_mask,
+            pad=(pad_size, pad_size),
+            value=mask_min_value,
+        )
+        return padded_hidden_states, padded_mask
+
+    @classmethod
+    def unpad_local_tokens(cls, padded_hidden_states, block_size):
+        # padded_hidden_states: [batch_size, padded seq_len, hidden_dim]
+        pad_size = block_size // 2
+        return padded_hidden_states[:, pad_size:-pad_size, :]
+
+
+class PegasusXDecoderLayer(nn.Module):
```

Comment:
Couldn't we add a `# Copied from ...` here from `PegasusDecoderLayer` ? What is the difference?

Reply:
Hmm, this is interesting.

Neither [PEGASUS](https://github.com/google-research/pegasus/blob/7530a6fe0d37c78dc23cbe42e68d62d1fb03f91c/pegasus/layers/attention.py#L41-L45) nor [PEGASUS-X](https://github.com/google-research/pegasus/blob/7530a6fe0d37c78dc23cbe42e68d62d1fb03f91c/pegasus/flax/models/decoders/extended/extended.py#L242) use bias terms in self-attention. However, it looks like BART does, and the current HF PEGASUS implementation does too (although it initializes them to 0, so that's not a problem).

The difference in the implementation here is the specification that `bias=False`. I actually don't feel that strongly here about adding zero-initialized bias terms by default, but that's the reason for the deviation from the PEGASUS implementation.